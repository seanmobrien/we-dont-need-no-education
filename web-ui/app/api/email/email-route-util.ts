/**
 * @module email-route-util
 *
 * This module provides utility functions shared by all email API routes.
 *
 * Functions:
 * - mapRecordToSummary: Maps a record to a summary object containing email details.
 * - mapRecordToObject: Maps a record to an object containing email details and body content.
 */

import { ContactSummary } from '@/data-models';
import { query, queryExt } from '@/lib/neondb';
import { ValidationError } from '@/lib/react-util';

/**
 * Maps a record to a summary object containing email details.
 *
 * @param {Record<string, unknown>} record - The record to map.
 * @returns {object} The mapped summary object.
 */
export const mapRecordToSummary = (record: Record<string, unknown>) => ({
  emailId: record.email_id,
  subject: record.subject,
  sentOn: record.sent_timestamp,
  threadId: record.thread_id,
  parentEmailId: record.parent_email_id,
  sender: {
    contactId: record.senderid,
    name: record.sendername,
    email: record.senderemail,
  },
  recipients: ((record.recipients as Array<Record<string, unknown>>) || []).map(
    (r: Record<string, unknown>) => ({
      contactId: r.recipient_id,
      name: r.recipient_name,
      email: r.recipient_email,
    })
  ),
});

/**
 * Maps a record to an object containing email details and body content.
 *
 * @param {Record<string, unknown>} record - The record to map.
 * @returns {object} The mapped object with email details and body content.
 */
export const mapRecordToObject = (record: Record<string, unknown>) => ({
  ...mapRecordToSummary(record),
  body: record.email_contents,
});

export const insertRecipients = async (
  emailId: number,
  recipients: ContactSummary[],
  clear: boolean = true,
  allowEmpty: boolean = false
) => {
  if (!allowEmpty && !recipients?.length) {
    throw new ValidationError({
      field: 'recipients',
      value: !recipients ? '[null]' : '[empty]',
      reason: 'At least one recipient is required',
      source: 'insertRecipients',
    });
  }

  // Delete existing recipients if clear is true
  if (clear) {
    await query(
      (sql) => sql`DELETE FROM email_recipients WHERE email_id = ${emailId}`
    );
  }
  if (!recipients.length) {
    return 0;
  }
  // Build bulk insert statement
  const insertSql = `INSERT INTO email_recipients (email_id, recipient_id)
    VALUES ${recipients
      .map((r: ContactSummary) => `(${emailId}, ${r.contactId})`)
      .join(', ')}
    RETURNING email_id, recipient_id
  `.toString();
  const res = await query((sql) => sql<false, false>(insertSql));
  if (res.length !== recipients.length) {
    throw new Error('Failed to insert all recipients');
  }
  return res.length;
};
