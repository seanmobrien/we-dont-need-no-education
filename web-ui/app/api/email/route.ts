import { NextRequest, NextResponse } from 'next/server';
import { query, queryExt } from '@/lib/neondb';
import { log } from '@/lib/logger';
import {
  normalizeNullableNumeric,
  parsePaginationStats,
} from '@/data-models/_utilities';
import {
  insertRecipients,
  mapRecordToObject,
  mapRecordToSummary,
} from '../../../lib/api/email/util';
import { LoggedError, ValidationError } from '@/lib/react-util';
import { buildOrderBy } from '@/lib/components/mui/data-grid/server';

/**
 * Handles the GET request to fetch a list of emails with sender and recipient information.
 *
 * This function queries the database to retrieve emails along with their sender's details and
 * a list of recipients. The results are returned as a JSON response.
 *
 * @returns {Promise<NextResponse>} A promise that resolves to a JSON response containing the
 * list of emails with sender and recipient information, or an error message if the request fails.
 *
 * @throws {Error} If there is an issue with the database query or any other error occurs during
 * the execution of the function, an error is logged and a 500 Internal Server Error response is returned.
 */
export async function GET(req: NextRequest): Promise<NextResponse> {
  try {
    const thisUrl = new URL(req.url);
    const { num, offset, page } = parsePaginationStats(thisUrl);
    // Fetch list of emails with sender info
    const result = await query(
      (sql) =>
        sql`SELECT 
          e.email_id,
          e.thread_id,
          e.subject,
          e.sent_timestamp,
          sender.contact_id AS senderId,
          sender.name AS senderName,
          sender.email AS senderEmail,
          (
            
              SELECT COALESCE(json_agg(json_build_object(
                          'recipient_id', recipient.contact_id,
                          'recipient_name', recipient.name,
                          'recipient_email', recipient.email
                      )
                ),'[]')						
                FROM email_recipients er 
                    LEFT JOIN contacts recipient ON er.recipient_id = recipient.contact_id 
                WHERE e.email_id = er.email_id AND recipient.contact_id IS NOT NULL    
          ) AS recipients,
          (SELECT COUNT(*) FROM email_attachments ea WHERE ea.email_id=e.email_id) AS count_attachments,
          (
            SELECT COUNT(*) FROM document_property dp 
            JOIN document_units d2 ON d2.unit_id=dp.document_id
            WHERE d2.email_id=e.email_id
              AND document_property_type_id=9
          ) AS count_kpi,
          (
            SELECT COUNT(*) FROM document_property dp 
            WHERE document_unit_email(dp.document_id)=e.email_id
              AND (document_property_type_id=102 OR document_property_type_id=1000)
          ) AS count_notes,
          (
            SELECT COUNT(*) FROM document_property dp 
            WHERE document_unit_email(dp.document_id)=e.email_id
              AND document_property_type_id=4
          ) AS count_cta,
          (
            SELECT COUNT(*) FROM document_property dp 
            WHERE document_unit_email(dp.document_id)=e.email_id
              AND document_property_type_id=5
          ) AS count_responsive_actions
        FROM emails e
        JOIN contacts sender ON e.sender_id = sender.contact_id                      
        ${buildOrderBy({ sql, source: req, defaultSort: [{ field: 'sent_timestamp', sort: 'desc' }], columnMap: { sentOn: 'sent_timestamp' } })}          
        LIMIT ${num} OFFSET ${offset};`,
      { transform: mapRecordToSummary },
    );
    log((l) =>
      l.verbose({ msg: '[[AUDIT]] -  Email list:', resultset: result, num, offset }),
    );
    const total = await query(
      (sql) => sql`SELECT COUNT(*) AS records  FROM emails e
        JOIN contacts sender ON e.sender_id = sender.contact_id`,
    );

    return NextResponse.json(
      {
        pageStats: { page, num, total: Number(total[0].records) },
        results: result,
      },
      { status: 200 },
    );
  } catch (error) {
    LoggedError.isTurtlesAllTheWayDownBaby(error, {
      log: true,
      source: 'GET email',
    });
    return NextResponse.json(
      { error: 'Internal Server Error' },
      { status: 500 },
    );
  }
}

/**
 * Handles the POST request to create a new email.
 *
 * @param {NextRequest} req - The incoming request object.
 * @returns {Promise<NextResponse>} - The response object containing the result of the email creation.
 * @throws {Error} - If there is an issue with the email creation process.
 */
export async function POST(req: NextRequest): Promise<NextResponse> {
  try {
    // Parse request email_contents
    const {
      senderId: incomingSenderId,
      subject,
      body,
      sentOn,
      threadId,
      parentEmailId,
      recipients,
      sender,
    } = await req.json();

    // Support taking in either a senderId or sender object, with precendence given to
    // the id bacause you have to go out of your way to set it.
    const senderId = incomingSenderId ?? sender?.contactId;

    // Validate required fields
    if (
      !senderId ||
      !subject ||
      !body ||
      !sentOn ||
      !recipients ||
      recipients.length === 0
    ) {
      return NextResponse.json(
        { error: 'Missing required fields' },
        { status: 400 },
      );
    }

    // Insert email into the database
    const result = await query(
      (
        sql,
      ) => sql`INSERT INTO emails (sender_id, subject, email_contents, sent_timestamp, thread_id, parent_email_id) 
       VALUES (${senderId}, ${subject}, ${body}, ${sentOn}, ${normalizeNullableNumeric(
         threadId,
       )}, ${normalizeNullableNumeric(parentEmailId)}) RETURNING *`,
      { transform: mapRecordToObject },
    );

    const emailId = Number(result[0].emailId);
    await insertRecipients(emailId, recipients);

    log((l) =>
      l.verbose({ msg: '[[AUDIT]] -  Email created:', resultset: result[0] }),
    );

    return NextResponse.json(
      {
        message: 'Email created successfully',
        email: result[0],
      },
      { status: 201 },
    );
  } catch (error) {
    if (ValidationError.isValidationError(error)) {
      return NextResponse.json({ error: error.message }, { status: 400 });
    }
    log((l) =>
      l.error({
        source: 'POST email',
        error,
      }),
    );
    return NextResponse.json(
      { error: 'Internal Server Error' },
      { status: 500 },
    );
  }
}

/**
 * Handles the PUT request to update an existing email.
 *
 * @param {NextRequest} req - The incoming request object.
 * @returns {Promise<NextResponse>} - The response object containing the result of the email update.
 * @throws {Error} - If there is an issue with the email update process.
 */
export async function PUT(req: NextRequest): Promise<NextResponse> {
  try {
    // Parse request email_contents
    const {
      emailId,
      senderId: incomingSenderId,
      subject,
      body,
      sentOn,
      threadId: incomingThreadId,
      parentEmailId: incomingParentEmailId,
      recipients,
      sender,
    } = await req.json();

    const threadId = normalizeNullableNumeric(incomingThreadId);
    const parentEmailId = normalizeNullableNumeric(incomingParentEmailId);
    // Support taking in either a senderId or sender object,
    // with precendence given ot the id bacause you have to go out of your way to set it.
    const senderId = incomingSenderId ?? sender?.contactId;

    // Validate required fields
    if (!emailId) {
      return NextResponse.json(
        { error: 'Email ID is required' },
        { status: 400 },
      );
    }

    // Validate that at least one field to update is provided
    if (
      !senderId &&
      !subject &&
      !body &&
      !sentOn &&
      (threadId ?? 0) < 1 &&
      (parentEmailId ?? 0) < 1
    ) {
      return NextResponse.json(
        { error: 'At least one field is required for update' },
        { status: 400 },
      );
    }

    // Build dynamic update query
    const updateFields: string[] = [];
    const values: unknown[] = [];
    let paramIndex = 1;

    if (senderId) {
      updateFields.push(`sender_id = $${paramIndex++}`);
      values.push(senderId);
    }
    if (subject) {
      updateFields.push(`subject = $${paramIndex++}`);
      values.push(subject);
    }
    if (body) {
      updateFields.push(`email_contents = $${paramIndex++}`);
      values.push(body);
    }
    if (sentOn) {
      updateFields.push(`sent_timestamp = $${paramIndex++}`);
      values.push(sentOn);
    }
    if (normalizeNullableNumeric(threadId)) {
      updateFields.push(`thread_id = $${paramIndex++}`);
      values.push(threadId);
    }
    if (normalizeNullableNumeric(parentEmailId)) {
      updateFields.push(`parent_email_id = $${paramIndex++}`);
      values.push(parentEmailId);
    }

    values.push(emailId); // Add email_id as the last parameter

    // Execute update query
    const result = await queryExt(
      (sql) =>
        sql<false, true>(
          `UPDATE emails SET ${updateFields.join(
            ', ',
          )} WHERE email_id = $${paramIndex} RETURNING *`,
          values,
        ),
      { transform: mapRecordToObject },
    );

    if (result.rowCount === 0) {
      return NextResponse.json(
        { error: 'Email not found or not updated' },
        { status: 404 },
      );
    }
    if (recipients && recipients.length > 0) {
      await insertRecipients(emailId, recipients, true);
    }

    log((l) =>
      l.verbose({ msg: '[[AUDIT]] -  Email updated:', resultset: result.rows[0] }),
    );

    return NextResponse.json(
      { message: 'Email updated successfully', email: result.rows[0] },
      { status: 200 },
    );
  } catch (error) {
    log((l) => l.error({ source: 'PUT email', error }));
    return NextResponse.json(
      { error: 'Internal Server Error' },
      { status: 500 },
    );
  }
}
