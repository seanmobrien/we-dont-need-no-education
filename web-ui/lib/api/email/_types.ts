import { PaginatedRequestApiParams } from '../_types';

/**
 * Parameters for the email search API.
 *
 * @extends PaginatedRequestApiParams
 *
 * @property {number[] | number} [contactId] - An optional contact ID or array of contact IDs to filter the search.
 * @property {string} [query] - **(Obsolete)** An optional search query. Use `q` instead.
 * @property {string} [q] - An optional search query.
 */
export type EmailSearchApiParams = PaginatedRequestApiParams & {
  contactId?: number[] | number;
  query?: string;
  q?: string;
};
