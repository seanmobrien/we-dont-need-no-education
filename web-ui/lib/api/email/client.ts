import type {
  EmailMessageSummary,
  EmailMessage,
  EmailMessageStats,
} from '@/data-models/api/email-message';
import type {
  PaginatedResultset,
  PaginationStats,
} from '@/data-models/_types';
import type { EmailSearchApiParams } from './types';
import siteMap from '@/lib/site-util/url-builder';
import {
  apiRequestHelperFactory,
  ApiRequestHelper,
} from '@/lib/send-api-request';
import { ICancellablePromiseExt } from '@/lib/typescript';

const apiRequest = <TResult>(
  cb: (api: ApiRequestHelper, builder: typeof siteMap.api.email) => TResult
): TResult => {
  const apiHelper = apiRequestHelperFactory({ area: 'email' });
  const builder = siteMap.api.email;
  return cb(apiHelper, builder);
};

/**
 * Fetches a list of email messages.
 *
 * @returns {ICancellablePromiseExt<ReadonlyArray<EmailMessageSummary>>} A ICancellablePromiseExt that resolves to an array of email message summaries.
 */
export const getEmailList = ({
  page,
  num,
}: Omit<PaginationStats, 'total'>): ICancellablePromiseExt<
  ReadonlyArray<EmailMessageSummary>
> =>
  apiRequest((apiHelper, builder) =>
    apiHelper.get<ReadonlyArray<EmailMessageSummary>>({
      url: builder.page({ page, num }),
      action: 'list',
    })
  );

/**
 * Fetches a specific email message by its ID.
 *
 * @param {number} id - The ID of the email message to fetch.
 * @returns {ICancellablePromiseExt<ReadonlyArray<EmailMessage>>} A ICancellablePromiseExt that resolves to the email message.
 */
export const getEmail = (id: string): ICancellablePromiseExt<EmailMessage> =>
  apiRequest((apiHelper, builder) =>
    apiHelper.get<EmailMessage>({
      url: builder.page(id),
      action: 'load',
    })
  );

/**
 * Creates a new email record.
 *
 * @param {Omit<EmailMessage, 'emailId'>} email - The email message data to create.
 * @returns {ICancellablePromiseExt<EmailMessage>} A ICancellablePromiseExt that resolves to the created email message.
 */
export const createEmailRecord = (
  email: Omit<EmailMessage, 'emailId'>
): ICancellablePromiseExt<EmailMessage> =>
  apiRequest((apiHelper, builder) =>
    apiHelper.post<EmailMessage>({
      url: builder.page(),
      action: 'create',
      input: email,
    })
  );

/**
 * Updates an existing email record.
 *
 * @param {EmailMessage} email - The email message data to update.
 * @returns {ICancellablePromiseExt<EmailMessage>} A ICancellablePromiseExt that resolves to the updated email message.
 */
export const updateEmailRecord = (
  email: EmailMessage
): ICancellablePromiseExt<EmailMessage> =>
  apiRequest((apiHelper, builder) =>
    apiHelper.put<EmailMessage>({
      url: builder.page(),
      action: 'update',
      input: email,
    })
  );

/**
 * Writes an email record. If the email has an `emailId` greater than 0, it updates the existing record.
 * Otherwise, it creates a new email record.
 *
 * @param email - The email message object. It omits the `emailId` property but allows it to be partially included.
 * @returns A cancellable promise that resolves to the email message.
 */
export const writeEmailRecord = (
  email: Omit<EmailMessage, 'emailId'> & Partial<Pick<EmailMessage, 'emailId'>>
): ICancellablePromiseExt<EmailMessage> =>
  email.emailId ?? 0 > 0
    ? updateEmailRecord(email as EmailMessage)
    : createEmailRecord(email);

/**
 * Deletes an email record by its ID.
 *
 * @param {number} id - The ID of the email message to delete.
 * @returns {ICancellablePromiseExt<EmailMessageSummary>} A ICancellablePromiseExt that resolves to the summary of the deleted email message.
 */
export const deleteEmailRecord = (
  id: number
): ICancellablePromiseExt<EmailMessageSummary> =>
  apiRequest((apiHelper, builder) =>
    apiHelper.delete<EmailMessageSummary>({
      url: builder.page(id),
      action: 'delete',
    })
  );

/**
 * Fetches email statistics.
 *
 * @returns {ICancellablePromiseExt<EmailMessageStats>} A ICancellablePromiseExt that resolves to the email message statistics.
 */
export const getEmailStats = (): ICancellablePromiseExt<EmailMessageStats> =>
  apiRequest((apiHelper, builder) =>
    apiHelper.get<EmailMessageStats>({
      url: builder.stats(),
      action: 'stats',
    })
  );

/**
 * Fetches email search results based on the provided search parameters.
 *
 * @param {EmailSearchApiParams} ops - The search parameters.
 * @returns {ICancellablePromiseExt<PaginatedResultset<EmailMessageSummary>>} A ICancellablePromiseExt that resolves to a paginated result set of email message summaries.
 */
export const getEmailSearchResults = (
  ops: EmailSearchApiParams
): ICancellablePromiseExt<PaginatedResultset<EmailMessageSummary>> =>
  apiRequest((apiHelper, builder) =>
    apiHelper.get<PaginatedResultset<EmailMessageSummary>>({
      url: builder.search(ops),
      action: 'search',
    })
  );
