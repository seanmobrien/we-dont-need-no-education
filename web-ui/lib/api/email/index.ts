export type * from './types';
export * from './database';
export * from './properties';
export * from './import';
export * from './util';
