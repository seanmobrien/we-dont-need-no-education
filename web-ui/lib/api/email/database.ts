import { ValidationError } from '@/lib/react-util/errors';
import { FirstParameter, newUuid } from '@/lib/typescript';
import { ObjectRepository } from '../_types';
import {
  EmailMessageAttachment,
  EmailMessageAttachmentSummary,
  EmailMessageSummary,
} from '@/data-models/api/email-message';
import { BaseObjectRepository } from '../_baseObjectRepository';
import { query } from '@/lib/neondb';
import { AbstractObjectRepository } from '../abstractObjectRepository';
import { db } from '@/lib/neondb';

type RepositoryEmailSummary = Omit<
  EmailMessageSummary,
  'sender' | 'recipients'
> & { senderId: number };

type RepositoryEmail = RepositoryEmailSummary & {
  emailContents: string;
  documentId?: number;
};

const mapRecordToSummary = (
  record: Record<string, unknown>,
): RepositoryEmailSummary => ({
  emailId: record.email_id as string,
  subject: record.subject as string,
  sentOn: record.sent_timestamp as Date,
  senderId: Number(record.sender_id),
  globalMessageId: (record.global_message_id as string) ?? undefined,
  parentEmailId: record.parent_email_id as string | null,
  importedFromId: record.imported_from_id
    ? (record.imported_from_id as string)
    : null,
  threadId: record.thread_id ? (record.thread_id as number) : undefined,
});

const mapRecordToObject = (record: Record<string, unknown>) => ({
  ...mapRecordToSummary(record),
  senderId: Number(record.sender_id),
  emailContents: record.email_contents as string,
});

export class EmailRepository extends BaseObjectRepository<
  RepositoryEmail,
  'emailId'
> {
  constructor() {
    super({
      tableName: 'emails',
      idField: 'emailId',
      objectMap: mapRecordToObject,
      summaryMap: mapRecordToSummary,
    });
  }

  public async create(
    props: Omit<RepositoryEmail, 'emailId'>,
  ): Promise<RepositoryEmail> {
    const ret = await super.create(props);
    if (ret && ret.emailId) {
      const importDate = props.sentOn ? new Date(props.sentOn) : new Date();
      const res = await db(
        (
          sql,
        ) => sql`INSERT INTO document_units (email_id, content, created_on, document_type)
        VALUES (${ret.emailId}, ${props.emailContents}, ${importDate.toISOString()}, 'email')
        returning unit_id`,
      );
      if (res.length < 1) {
        throw new Error('Failed to create document unit');
      }
      ret.documentId = res[0].unit_id;
    }
    return ret;
  }
  /**
   * Validates the input for a specific method.
   *
   * @template TMethod
   * @param {TMethod} method - The method to validate.
   * @param {FirstParameter<ObjectRepository<T, KId>[TMethod]>} obj - The input to validate.
   */
  validate<TMethod extends keyof ObjectRepository<RepositoryEmail, 'emailId'>>(
    method: TMethod,
    obj: FirstParameter<
      Pick<ObjectRepository<RepositoryEmail, 'emailId'>, TMethod>[TMethod]
    >,
  ): void {
    const asModel = obj as RepositoryEmail;
    switch (method) {
      case 'create':
        if (!asModel.senderId || !asModel.subject || !asModel.emailContents) {
          throw new ValidationError({
            field: 'senderId||subject||emailContents',
            source: 'EmailRepository',
          });
        }
        break;
      case 'update':
        if (
          !asModel.emailId ||
          (!asModel.senderId &&
            !asModel.subject &&
            !asModel.globalMessageId &&
            !asModel.importedFromId &&
            !asModel.emailContents &&
            !asModel.sentOn)
        ) {
          throw new ValidationError({
            field: 'emailId||At least one field is required for update',
            source: 'EmailRepository',
          });
        }
        break;
      default:
        break;
    }
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected getListQueryProperties(): [string, Array<any>, string] {
    return [
      `SELECT * FROM emails ORDER BY sent_timestamp`,
      [],
      `SELECT COUNT(*) as records FROM emails`,
    ];
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  protected getQueryProperties(emailId: string): [string, Array<any>] {
    return ['SELECT * FROM emails WHERE email_id = $1', [emailId]];
  }
  protected getCreateQueryProperties(
    obj: RepositoryEmail,
  ): [
    string,
    [
      string,
      number,
      number | null,
      string | null,
      string,
      string,
      string | null,
      string | null,
      string | Date | null,
    ],
  ] {
    if (!obj.emailId) {
      obj.emailId = newUuid();
    }
    return [
      'INSERT INTO emails (email_id, sender_id, thread_id, parent_id, subject, email_contents, global_message_id, imported_from_id, sent_timestamp) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING email_id',
      [
        obj.emailId,
        obj.senderId,
        obj.threadId ?? null,
        obj.parentEmailId ?? null,
        obj.subject,
        obj.emailContents,
        obj.globalMessageId ?? null,
        obj.importedFromId ?? null,
        !obj.sentOn ? null : typeof obj.sentOn === 'string' ? obj.sentOn : obj.sentOn.toISOString(),
      ],
    ];
  }
  protected getUpdateQueryProperties(
    obj: RepositoryEmail,
  ): [Record<string, unknown>] {
    return [
      {
        sender_id: obj.senderId,
        thread_id: obj.threadId,
        parent_email_id: obj.parentEmailId,
        subject: obj.subject,
        email_contents: obj.emailContents,
        sent_timestamp: obj.sentOn,
        imported_from_id: obj.importedFromId,
        global_message_id: obj.globalMessageId,
      },
    ];
  }
  async getIdForUniqueMessageId(
    uniqueId: string | null,
  ): Promise<string | null> {
    if (!uniqueId) return null;
    try {
      const result = await query(
        (sql) =>
          sql`select email_id from emails where global_message_id = ${uniqueId}`,
      );
      return result.length === 1 ? (result[0].email_id as string) : null;
    } catch (error) {
      AbstractObjectRepository.logDatabaseError(this.source, error);
    }
    return null;
  }
}

const mapAttachmentRecordToSummary = (
  record: Record<string, unknown>,
): EmailMessageAttachmentSummary => ({
  attachmentId: record.attachment_id as number,
  emailId: record.email_id as string,
  fileName: record.file_name as string,
  filePath: record.file_path as string,
  size: record.size as number,
  mimeType: record.mime_type as string,
});

const mapAttachmentRecordToObject = (
  record: Record<string, unknown>,
): EmailMessageAttachment => ({
  ...mapAttachmentRecordToSummary(record),
  extractedText: record.extracted_text as string | null,
  extractedTextVector: record.extracted_text_vector as string | null,
  policyId: record.policy_id as number | null,
  summary: record.summary as string | null,
});

export class EmailAttachmentRepository extends BaseObjectRepository<
  EmailMessageAttachment,
  'attachmentId'
> {
  constructor() {
    super({
      tableName: 'email_attachments',
      idField: 'attachmentId',
      objectMap: mapAttachmentRecordToObject,
      summaryMap: mapAttachmentRecordToSummary,
    });
  }

  /**
   * Validates the input for a specific method.
   *
   * @template TMethod
   * @param {TMethod} method - The method to validate.
   * @param {FirstParameter<ObjectRepository<T, KId>[TMethod]>} obj - The input to validate.
   */
  validate<
    TMethod extends keyof ObjectRepository<
      EmailMessageAttachment,
      'attachmentId'
    >,
  >(
    method: TMethod,
    obj: FirstParameter<
      Pick<
        ObjectRepository<EmailMessageAttachment, 'attachmentId'>,
        TMethod
      >[TMethod]
    >,
  ): void {
    const asModel = obj as EmailMessageAttachment;
    switch (method) {
      case 'create':
        if (!asModel.fileName || !asModel.filePath || !asModel.emailId) {
          throw new ValidationError({
            field: 'fileName||filePath||emailId',
            source: 'EmailAttachmentRepository',
          });
        }
        break;
      case 'update':
        if (
          !asModel.attachmentId ||
          (!asModel.fileName &&
            !asModel.filePath &&
            !asModel.emailId &&
            !asModel.extractedText &&
            !asModel.extractedTextVector &&
            !asModel.size &&
            !asModel.mimeType &&
            !asModel.policyId &&
            !asModel.summary)
        ) {
          throw new ValidationError({
            field: 'attachmentId||At least one field is required for update',
            source: 'EmailAttachmentRepository',
          });
        }
        break;
      default:
        break;
    }
  }
  protected getQueryProperties(attachmentId: number): [string, Array<number>] {
    return [
      'SELECT * FROM email_attachments WHERE attachment_id = $1',
      [attachmentId],
    ];
  }
  protected getCreateQueryProperties(
    obj: EmailMessageAttachment,
  ): [
    string,
    [
      string,
      string,
      string | null,
      string | null,
      number | null,
      string | null,
      string,
      number,
      string,
    ],
  ] {
    return [
      'INSERT INTO email_attachments (file_name, file_path, extracted_text, extracted_text_tsv, policy_id, summary, email_id, size, mime_type) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9) RETURNING email_id',
      [
        obj.fileName,
        obj.filePath,
        obj.extractedText,
        obj.extractedTextVector,
        obj.policyId,
        obj.summary,
        obj.emailId,
        obj.size,
        obj.mimeType,
      ],
    ];
  }
  protected getUpdateQueryProperties(
    obj: EmailMessageAttachment,
  ): [Record<string, string | number | null>] {
    return [
      {
        file_name: obj.fileName,
        file_path: obj.filePath,
        extracted_text: obj.extractedText,
        extracted_text_tsv: obj.extractedTextVector,
        policy_id: obj.policyId,
        summary: obj.summary,
        email_id: obj.emailId,
        size: obj.size,
        mime_type: obj.mimeType,
      },
    ];
  }
}
