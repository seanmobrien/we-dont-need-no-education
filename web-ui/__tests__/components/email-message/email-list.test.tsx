import { render, screen, waitFor } from '@testing-library/react';
import EmailList from 'components/email-message/email-list';
import { mockEmail, mockEmailSummary } from '../email.mock-data';

const testIds = {
  noneFound: 'email-list-none-found',
  error: 'email-list-error',
  sender: (emailId: number) => `email-list-sender-${emailId}`,
  subject: (emailId: number) => `email-list-subject-${emailId}`,
  timestamp: (emailId: number) => `email-list-timestamp-${emailId}`,
  editHeader: 'email-list-edit-header',
};

describe('EmailList', () => {
  const mockFetch = jest.fn();

  beforeAll(() => {
    global.fetch = mockFetch;
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should display loading state initially', () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => [],
    });

    const { asFragment } = render(<EmailList />);
    const theElement = asFragment();
    expect(theElement).toMatchSnapshot();
  });

  it('should display error message when fetching emails fails', async () => {
    mockFetch.mockRejectedValueOnce(new Error('Error fetching emails.'));

    render(<EmailList />);

    await waitFor(() => {
      const error = expect(screen.getByTestId(testIds.error));
      error.toBeInTheDocument();
      error.toHaveTextContent('Error fetching emails.');
    });
  });

  it('should display no emails found message when there are no emails', async () => {
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => [],
    });

    render(<EmailList />);

    await waitFor(() => {
      expect(screen.getByTestId(testIds.noneFound)).toBeInTheDocument();
    });
  });

  it('should display a list of emails', async () => {
    const mockEmails = mockEmailSummary();
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => mockEmails,
    });
    const { asFragment } = render(<EmailList />);
    await waitFor(() => {
      expect(screen.getByTestId(testIds.sender(1))).toBeInTheDocument();
      expect(screen.getByTestId(testIds.subject(1))).toBeInTheDocument();
      expect(screen.getByTestId(testIds.sender(2))).toBeInTheDocument();
      expect(screen.getByTestId(testIds.subject(2))).toBeInTheDocument();
    });

    const theElement = asFragment();
    expect(theElement).toMatchSnapshot();
  }, 5000);

  it('should display the email form when an email is selected', async () => {
    const mockEmails = mockEmailSummary();
    mockFetch.mockResolvedValueOnce({
      ok: true,
      json: async () => mockEmails,
    });
    render(<EmailList />);
    await waitFor(
      () => {
        const subjectNode = screen.getByTestId(testIds.subject(1));
        expect(subjectNode).toBeInTheDocument();

        mockFetch.mockResolvedValueOnce({
          ok: true,
          json: async () => mockEmail(),
        });
        subjectNode.click();
        expect(screen.getByTestId(testIds.editHeader)).toBeInTheDocument();
      },
      { timeout: 10000 }
    );
  });
});
