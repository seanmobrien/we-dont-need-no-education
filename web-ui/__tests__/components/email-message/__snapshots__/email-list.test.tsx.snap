// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailList should display a list of emails 1`] = `
<DocumentFragment>
  <div
    class="max-w-3xl mx-auto p-6 bg-white rounded-lg shadow-md"
  >
    <h2
      class="text-xl font-semibold mb-4"
    >
      Email List
    </h2>
    <div>
      <div
        class="flex items-center justify-between p-4 mb-2 border rounded bg-gray-50 hover:bg-gray-100 cursor-pointer transition hover:shadow"
      >
        <div>
          <p
            class="font-semibold text-gray-800"
          >
            <a
              data-testid="email-list-sender-1"
              href="#"
            >
              John Smith
            </a>
          </p>
          <p
            class="text-gray-600"
          >
            <a
              data-testid="email-list-subject-1"
              href="#"
            >
              Test Subject 1
            </a>
          </p>
        </div>
        <p
          class="text-sm text-gray-600"
          data-testid="email-list-timestamp-1"
        >
          12/31/2022, 6:00:00 PM
        </p>
      </div>
      <div
        class="flex items-center justify-between p-4 mb-2 border rounded bg-gray-50 hover:bg-gray-100 cursor-pointer transition hover:shadow"
      >
        <div>
          <p
            class="font-semibold text-gray-800"
          >
            <a
              data-testid="email-list-sender-2"
              href="#"
            >
              Jane Doe
            </a>
          </p>
          <p
            class="text-gray-600"
          >
            <a
              data-testid="email-list-subject-2"
              href="#"
            >
              Test Subject 2
            </a>
          </p>
        </div>
        <p
          class="text-sm text-gray-600"
          data-testid="email-list-timestamp-2"
        >
          1/2/2023, 12:00:00 AM
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`EmailList should display loading state initially 1`] = `
<DocumentFragment>
  <div
    class="max-w-3xl mx-auto p-6 bg-white rounded-lg shadow-md"
  >
    <h2
      class="text-xl font-semibold mb-4"
    >
      Email List
    </h2>
    <p
      class="text-gray-600"
    >
      Loading emails...
    </p>
  </div>
</DocumentFragment>
`;
